File: examples/go/xauth.go
package xauth

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"runtime"
	"strings"
)

func getHWID() string {
	switch runtime.GOOS {
	case "windows":
		out, err := exec.Command("wmic", "csproduct", "get", "uuid").Output()
		if err == nil {
			lines := strings.Split(string(out), "\n")
			if len(lines) > 1 {
				return strings.TrimSpace(lines[1])
			}
		}
	case "linux":
		out, err := os.ReadFile("/etc/machine-id")
		if err == nil {
			return strings.TrimSpace(string(out))
		}
	case "darwin":
		out, err := exec.Command("sh", "-c", "ioreg -rd1 -c IOPlatformExpertDevice | awk '/IOPlatformUUID/ { print $3; }'").Output()
		if err == nil {
			return strings.TrimSpace(strings.ReplaceAll(string(out), "\"", ""))
		}
	}
	
	// Fallback
	host, _ := os.Hostname()
	return fmt.Sprintf("%x", sha256.Sum256([]byte(host)))
}

type XAuth struct {
	AppID     int
	AppSecret string
	BaseURL   string
	sessionID string
	nonce     string
}

type ValidationResult struct {
	Success   bool                   `json:"success"`
	Message   string                 `json:"message,omitempty"`
	Expiry    string                 `json:"expiry,omitempty"`
	Broadcast string                 `json:"broadcast,omitempty"`
	Variables map[string]interface{} `json:"variables,omitempty"`
}

func NewXAuth(appID int, appSecret string, baseURL string) *XAuth {
	if baseURL == "" {
		baseURL = "http://localhost:3310"
	}
	return &XAuth{
		AppID:     appID,
		AppSecret: appSecret,
		BaseURL:   strings.TrimRight(baseURL, "/"),
	}
}

func (x *XAuth) Initialize() error {
	payload, _ := json.Marshal(map[string]int{"app_id": x.AppID})
	
	resp, err := http.Post(x.BaseURL+"/api/v1/client/initialize", "application/json", bytes.NewBuffer(payload))
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return fmt.Errorf("initialization failed with status: %d", resp.StatusCode)
	}

	var data map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return err
	}

	if sid, ok := data["session_id"].(string); ok {
		x.sessionID = sid
	}
	if n, ok := data["nonce"].(string); ok {
		x.nonce = n
	}

	return nil
}

func (x *XAuth) decryptResponse(encryptedData string) (map[string]interface{}, error) {
	parts := strings.Split(encryptedData, ":")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid format")
	}

	iv, _ := base64.StdEncoding.DecodeString(parts[0])
	tag, _ := base64.StdEncoding.DecodeString(parts[1])
	ciphertext, _ := base64.StdEncoding.DecodeString(parts[2])

	rawKey := []byte(x.AppSecret + x.nonce)
	hash := sha256.Sum256(rawKey)

	block, err := aes.NewCipher(hash[:])
	if err != nil {
		return nil, err
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	// GCM expects ciphertext + tag
	ciphertextWithTag := append(ciphertext, tag...)
	plaintext, err := aesGCM.Open(nil, iv, ciphertextWithTag, nil)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(plaintext, &result); err != nil {
		return nil, err
	}

	return result, nil
}

func (x *XAuth) ValidateLicense(licenseKey string) *ValidationResult {
	if x.sessionID == "" || x.nonce == "" {
		if err := x.Initialize(); err != nil {
			return &ValidationResult{Success: false, Message: "Initialization failed: " + err.Error()}
		}
	}

	payload, _ := json.Marshal(map[string]interface{}{
		"license_key": licenseKey,
		"hwid":        getHWID(),
		"app_secret":  x.AppSecret,
		"session_id":  x.sessionID,
	})

	resp, err := http.Post(x.BaseURL+"/api/v1/client/validate", "application/json", bytes.NewBuffer(payload))
	if err != nil {
		return &ValidationResult{Success: false, Message: "Network error: " + err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode == 200 {
		var data map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&data)
		
		if encryptedPayload, ok := data["data"].(string); ok {
			decrypted, err := x.decryptResponse(encryptedPayload)
			if err == nil {
				if status, ok := decrypted["status"].(string); ok && status == "success" {
					res := &ValidationResult{Success: true}
					if expiry, ok := decrypted["expiry"].(string); ok {
						res.Expiry = expiry
					}
					if broadcast, ok := decrypted["broadcast"].(string); ok {
						res.Broadcast = broadcast
					}
					if variables, ok := decrypted["variables"].(map[string]interface{}); ok {
						res.Variables = variables
					}
					return res
				}
			}
		}
		return &ValidationResult{Success: false, Message: "Failed to verify secure response."}
	}

	bodyBytes, _ := io.ReadAll(resp.Body)
	var errData map[string]interface{}
	if err := json.Unmarshal(bodyBytes, &errData); err == nil {
		if msg, ok := errData["message"].(string); ok {
			return &ValidationResult{Success: false, Message: msg}
		}
	}
	return &ValidationResult{Success: false, Message: string(bodyBytes)}
}

File: examples/php/xauth.php
<?php

/**
 * XAuth Client SDK for PHP
 */
class XAuth {
    private $appId;
    private $appSecret;
    private $baseUrl;
    private $sessionId;
    private $nonce;

    public function __construct($appId, $appSecret, $baseUrl = "http://localhost:3310") {
        $this->appId = $appId;
        $this->appSecret = $appSecret;
        $this->baseUrl = rtrim($baseUrl, '/');
    }

    private function getHwid() {
        if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
            @exec('wmic csproduct get uuid', $output);
            if (isset($output[1])) return trim($output[1]);
        } elseif (PHP_OS === 'Linux') {
            if (file_exists('/etc/machine-id')) {
                return trim(file_get_contents('/etc/machine-id'));
            }
        } elseif (PHP_OS === 'Darwin') {
            @exec("ioreg -rd1 -c IOPlatformExpertDevice | awk '/IOPlatformUUID/ { print $3; }'", $output);
            if (isset($output[0])) return trim(str_replace('"', '', $output[0]));
        }
        return hash('sha256', gethostname() . get_current_user());
    }

    private function request($endpoint, $payload) {
        $ch = curl_init($this->baseUrl . $endpoint);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/json']);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($payload));
        
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        
        return ['status' => $httpCode, 'body' => $response];
    }

    public function initialize() {
        $res = $this->request("/api/v1/client/initialize", ["app_id" => $this->appId]);
        
        if ($res['status'] == 200) {
            $data = json_decode($res['body'], true);
            $this->sessionId = $data['session_id'];
            $this->nonce = $data['nonce'];
            return ['success' => true];
        }
        return ['success' => false, 'message' => $res['body']];
    }

    private function decryptResponse($encryptedData) {
        try {
            $parts = explode(":", $encryptedData);
            if (count($parts) != 3) return null;
            
            $iv = base64_decode($parts[0]);
            $tag = base64_decode($parts[1]);
            $ciphertext = base64_decode($parts[2]);

            $rawKey = $this->appSecret . $this->nonce;
            $key = hash('sha256', $rawKey, true);

            $decrypted = openssl_decrypt(
                $ciphertext,
                'aes-256-gcm',
                $key,
                OPENSSL_RAW_DATA,
                $iv,
                $tag
            );

            if ($decrypted === false) return null;
            return json_decode($decrypted, true);
        } catch (Exception $e) {
            return null;
        }
    }

    public function validateLicense($licenseKey) {
        if (!$this->sessionId || !$this->nonce) {
            $init = $this->initialize();
            if (!$init['success']) {
                return ['success' => false, 'message' => 'Initialization failed: ' . $init['message']];
            }
        }

        $res = $this->request("/api/v1/client/validate", [
            "license_key" => $licenseKey,
            "hwid" => $this->getHwid(),
            "app_secret" => $this->appSecret,
            "session_id" => $this->sessionId
        ]);

        if ($res['status'] == 200) {
            $data = json_decode($res['body'], true);
            $decryptedData = $this->decryptResponse($data['data']);
            
            if ($decryptedData && isset($decryptedData['status']) && $decryptedData['status'] === 'success') {
                return [
                    'success' => true,
                    'expiry' => $decryptedData['expiry'] ?? null,
                    'broadcast' => $decryptedData['broadcast'] ?? null,
                    'variables' => $decryptedData['variables'] ?? []
                ];
            }
            return ['success' => false, 'message' => 'Failed to verify secure response.'];
        } else {
            $errData = json_decode($res['body'], true);
            return [
                'success' => false, 
                'message' => $errData['message'] ?? 'Validation rejected'
            ];
        }
    }

    public function checkVersion($currentVersion = null, $channel = "stable") {
        $res = $this->request("/api/v1/client/verify-version", [
            "app_id" => $this->appId,
            "app_secret" => $this->appSecret,
            "channel" => $channel,
            "current_version" => $currentVersion
        ]);

        if ($res['status'] == 200) {
            $data = json_decode($res['body'], true);
            return [
                'success' => true,
                'version' => $data['version'] ?? null,
                'channel' => $data['channel'] ?? null,
                'url' => $data['url'] ?? null,
                'checksum' => $data['checksum'] ?? null,
                'published_at' => $data['published_at'] ?? null,
                'update_available' => $data['update_available'] ?? false,
                'broadcast' => $data['broadcast'] ?? null
            ];
        } else {
            $errData = json_decode($res['body'], true);
            return [
                'success' => false, 
                'message' => $errData['message'] ?? 'Version check failed'
            ];
        }
    }

    public function getLatestRelease($channel = "stable") {
        $res = $this->request("/api/update/{$this->appId}/{$channel}", null);
        // Note: request method uses POST by default, we need to handle GET for this specific route
        
        $ch = curl_init($this->baseUrl . "/api/update/{$this->appId}/{$channel}");
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

        if ($httpCode == 200) {
            $data = json_decode($response, true);
            return [
                'success' => true,
                'version' => $data['version'] ?? null,
                'channel' => $data['channel'] ?? null,
                'url' => $data['url'] ?? null,
                'checksum' => $data['checksum'] ?? null,
                'published_at' => $data['published_at'] ?? null
            ];
        }
        return ['success' => false, 'message' => $response];
    }
}

// Example Usage
/*
$auth = new XAuth(1, "your_app_secret_here");

echo "--- XAuth Omega PHP SDK Demo ---\n";

// 1. Public Update
echo "[1] Public Update Check: ";
$release = $auth->getLatestRelease();
echo $release['success'] ? "Latest: " . $release['version'] : "Error: " . $release['message'];
echo "\n";

// 2. Secure Check
echo "[2] Secure Version Verification: ";
$ver = $auth->checkVersion("1.0.0");
echo $ver['success'] ? ($ver['update_available'] ? "Update found!" : "Up to date.") : "Error";
echo "\n";

// 3. Validation
echo "[3] License Validation: ";
$result = $auth->validateLicense("XXXX-XXXX-XXXX-XXXX");
echo $result['success'] ? "Valid!" : "Invalid: " . $result['message'];
echo "\n";
*/
